team.py

```diff
@@ -52,46 +65,124 @@ def edit_password(self, teamname, input):
             except Landmark.DoesNotExist:
-                return "Congrats! You win!"
+                team.game_ended = True
+                if self.check_first(teamname):
+                    team.score += 50
+                team.save()
+                if self.check_game_over():
+                    game.running = False
+                    game.save()
+                return "Congrats! You've reached the end of the game!"
             team.current_landmark = next_landmark
+            team.question_requested = False
             team.save()
             ret_string = "Correct answer given! You can now request the clue for the next landmark"
         return ret_string
 
     def request_clue(self, teamname):
         # teamname: string
+        time_check = self.check_time_up()
         game = Game.objects.get(name="game")
         if not game.running:
+            if time_check:
+                return "Time is up!"
             return "There is no game running!"
         try:
             team = HuntUser.objects.get(name=teamname)
         except HuntUser.DoesNotExist:
             return "Not a valid team!"
-        if team.current_landmark.order_num == -1:
+        if team.game_ended:
+            return "You've already finished the game!"
+        if team.current_landmark.name == "dummy":
             return "Not at a valid landmark"
         return team.current_landmark.clue
 
     def request_question(self, teamname):
         # teamname: string
+        time_check = self.check_time_up()
         game = Game.objects.get(name="game")
         if not game.running:
+            if time_check:
+                return "Time is up!"
             return "There is no game running!"
         try:
             team = HuntUser.objects.get(name=teamname)
         except HuntUser.DoesNotExist:
             return "Not a valid team!"
+        if team.game_ended:
+            return "You've already finished the game!"
+        if team.current_landmark.name == "dummy":
+            return "Not at a valid landmark"
+        if not team.question_requested:
+            team.time_requested = datetime.now(tz=timezone.utc)
+            team.question_requested = True
+        team.save()
         return team.current_landmark.question
+
+    def check_first(self, teamname):
+        teams = HuntUser.objects.all()
+        for team in teams:
+            if team.name == "maker" or team.name == teamname:
+                continue
+            if team.game_ended:
+                return False
+        return True
+
+    def check_game_over(self):
+        teams = HuntUser.objects.all()
+        for team in teams:
+            if team.name == "maker":
+                continue
+            if not team.game_ended:
+                return False
+        return True
+
+    def check_time_up(self):
+        # First, check if there is a game running. If not, return False
```

Comment:
Bit confused by this class. I thought the only time we check is how long it takes to answer a question; why are we checking the total game time?